### PR TITLE
win: Allow sockets to be bound prior to uv_tcp_open/uv_listen.

### PR DESCRIPTION
**Pull request by 1st1**
I'm working on porting [uvloop](https://github.com/MagicStack/uvloop) to Windows. One of the unittests that works on Unix and doesn't on Windows is creating a server (with `asyncio.loop.create_server()`) out of an existing bound Python socket.

On Unix that works: you can pass a bound socket to `uv_tcp_open()`, and `uv_listen()` doesn't try to bind it.

On Windows, it's expected that you will call `uv_tcp_bind()` after `uv_tcp_open()`, or otherwise `uv_listen()` **will bind** the socket.

To check if socket is already bound in `uv_listen()`, we can call `getsockname()`, which on Windows errors out with `WSAEINVAL` for unbound sockets. If we have an error, it means that the socket is bound and we don't need to `bind()` it again.

### CODE REVIEW

File: src/win/tcp.c [hunk 1]
```diff
@@ -575,14 +577,20 @@ int uv_tcp_listen(uv_tcp_t* handle, int backlog, uv_connection_cb cb) {
...
   if (!(handle->flags & UV_HANDLE_BOUND)) {
-    err = uv_tcp_try_bind(handle,
-                          (const struct sockaddr*) &uv_addr_ip4_any_,
-                          sizeof(uv_addr_ip4_any_),
-                          0);
-    if (err)
-      return err;
-    if (handle->delayed_error)
-      return handle->delayed_error;
+    if (!uv_tcp_getsockname(handle, &saddr, &saddr_len)) {
```
Comment (saghul) [inline src/win/tcp.c]: I think this is more appropriate to do in `uv_tcp_open`, not here.

File: src/win/tcp.c [hunk 2]
```diff
@@ -556,6 +556,8 @@ static void uv_tcp_queue_read(uv_loop_t* loop, uv_tcp_t* handle) {
 
 int uv_tcp_listen(uv_tcp_t* handle, int backlog, uv_connection_cb cb) {
   uv_loop_t* loop = handle->loop;
+  struct sockaddr saddr;
```
Comment (saghul) [inline src/win/tcp.c]: This won't work for IPv6 addresses. Use `struct sockaddr_storage` and cast it to `struct sockaddr`.

**Comment by saghul** [comment]
Ping?

**Comment by saghul** [comment]
(When it lands) this patch would also fix https://github.com/libuv/libuv/issues/397

**Comment by 1st1** [comment]
Hi, I plan to get back to this (and the other PR) in a few days, but let me start a discussion now.  

With this PR I can implement `asyncio.loop.create_server` API, but when I started to port `asyncio.loop.connect_accepted_socket` and `asyncio.loop.create_connection` I hit another roadblock.The problem with the current `uv_tcp_open` function is that it could be used on *nix to work with any kind of socket: connected, accepted, and listened on.

On Windows the story is a bit different, because before starting to work with the socket you have to call `setsockopt(SOL_SOCKET, SO_UPDATE_ACCEPT_CONTEXT...)` or `setsockopt(SOL_SOCKET, SO_UPDATE_CONNECT_CONTEXT...)` depending on where your socket came from.

So detecting if the socket is already bound (what this PR is currently doing) doesn't solve all problems. We also need to find a way to communicate to libuv to call the correct `setsockopt` for sockets passed to `uv_tcp_open`.

One way to solve this problem is to add a new API: `uv_tcp_open2`:

```c
int uv_tcp_open2(uv_tcp_t* handle, uv_os_sock_t sock, unsigned int flags)
```

where flags will be defined as:

```c
UV_SOCK_BOUND        // the socket is already bound and ready to accept new connections
UV_SOCK_CONNECTED    // the socket is already connected, set SO_UPDATE_CONNECT_CONTEXT 
UV_SOCK_ACCEPTED     // the socket is already connected, set SO_UPDATE_ACCEPT_CONTEXT
```

The function will only do something on Windows, on *nix it will simply call `uv_tcp_open`.

Thoughts?

**Comment by saghul** [comment]
> On Windows the story is a bit different, because before starting to work with the socket you have to call setsockopt(SOL_SOCKET, SO_UPDATE_ACCEPT_CONTEXT...) or setsockopt(SOL_SOCKET, SO_UPDATE_CONNECT_CONTEXT...) depending on where your socket came from.

Why do we need this?  What doesn't work if we don't do it?

**Comment by 1st1** [comment]
Here are a couple of links:

* http://stackoverflow.com/questions/9169086/tcp-shutdown-with-sockets-connected-through-acceptex

* https://msdn.microsoft.com/en-us/library/windows/desktop/ms740532.aspx (search for `SO_UPDATE_ACCEPT_CONTEXT`)

I'm not an expert in Windows IOCP, but it seems that we have to do this for `getpeername`, `getsockname`, `getsockopt`, or `setsockopt` to work with the socket correctly. And that's what libuv is already doing, just not with `uv_tcp_open`.

**Comment by saghul** [comment]
Ah, I see.

One problem I see is that for the accepted sockets you need to call the function with the original listening socket.

Given the above, I'd say the application should call those for us before passing the fd.  Then we use getsockname and initialize the needful in `uv_tcp_open`.  It's a bit more work for the user, yes, but they are already in platform dependent territory if using this API, and I'd rather not add a new API to libuv because it would be Windows only.

**Comment by bnoordhuis** [comment]
@1st1 I'll close this out in favor of #1447.  If you want to keep working on this, let me know and I'll reopen.

**Pull request closed without merging**